### PR TITLE
handle worker crash

### PR DESCRIPTION
**Pull request by xoviat**
@sallner What do you think about this?

### CODE REVIEW

**Comment by xoviat** [comment]
Potentially, we can avoid using a database using the `pytest_report_from_serializable` hook in the controller to count failures, and re-configuring the failure history for tests. However, the number of reruns needs to be communicated to the controller before the test run is started. Otherwise, the controller will not be able to determine whether the crash should be re-run or not.

About the db approach: it's as good as it could be, but for the fact that it doesn't work when the nodes are executed on a remote server (as is possible with xdist).

**Comment by icemac** [comment]
This PR fails in all configurations, could you please have a look.

**Comment by xoviat** [comment]
The reason for the failure is that it uses hooks from pytest-xdist.

**Comment by icemac** [comment]
> The reason for the failure is that it uses hooks from pytest-xdist.

So a release of `pytest-xdist` is needed? But on the other hand `pytest-xdist` is not a dependency for `pytest-rerunfailures` and I do not think that it should become one.

**Comment by xoviat** [comment]
I still plan to complete this, but it will be delayed for a bit. The main problem is that the xdist controller needs to know whether a worker crash should be reported as a rerun or a failure, which requires the workers to communicate this information to the controller before running the test setup (so that if the test setup crashes, the controller can rerun it). Currently this is done through a file, but that appears to only work on windows and it doesn't work if the workers are remote. The new plan is set-up a simple key-value database on the controller that's stored in-memory.

xdist will not be a dependency in the final PR, but this functionality will be included if xdist is detected.

**Comment by icemac** [comment]
@xoviat Thank you for the clarification.

**Comment by xoviat** [comment]
new plan: start socket server on controller, use as db:

https://stackoverflow.com/questions/7749341/basic-python-client-socket-example

**Comment by xoviat** [comment]
@icemac I believe this may be close to complete. Do you have any questions about this, or should I clean up the commits?

**Comment by icemac** [comment]
@xoviat The last CI run was a crash. Each and every test failed and had to be killed by GHA after 6 hours. (This does not increase my confidence in this PR.)
What went wrong there and how are you planning to fix it?

**Comment by xoviat** [comment]
The latest run passed (https://github.com/xoviat/pytest-rerunfailures/actions/runs/903077274)

2bff11b wasn't ready yet, I was just pushing changes so that I could see the diffs on github.

File: .github/workflows/test.yml
Comment (icemac) [inline .github/workflows/test.yml:5]: This change is not needed. To run the PR in this repository (not the forks) I have to approve the run. This was introduced by GitHub to prevent crypto mining.

File: pytest_rerunfailures.py
Comment (icemac) [inline pytest_rerunfailures.py:271]: According to the code of `StatusDB` it is not actually a `no-op` – why?

File: pytest_rerunfailures.py
Comment (icemac) [inline pytest_rerunfailures.py:378]: I do not understand the meaning of the parameters, maybe longer variable names could clarify.

**Review by icemac — COMMENTED** [review]
I have some questions.

**Comment by icemac** [comment]
Additionally the linter tests fail.

There is one skipped test: Could it be that this is your newly written test?

File: .github/workflows/test.yml
Comment (xoviat) [reply .github/workflows/test.yml:5]: I will revert it. It was used to run it on my own fork (couldn't test it on this one because of the approvals).

File: pytest_rerunfailures.py
Comment (xoviat) [reply pytest_rerunfailures.py:271]: If xdist is not detected, then we don't need to record failures across workers, because a test crash `os.exit(1)` will crash pytest. Therefore, this functionality is irrelevant without xdist. With xdist, we need to record failures across workers because we need to reduce the number of times to rerun a test by the number of times that test has crashed a worker.

File: pytest_rerunfailures.py
Comment (xoviat) [reply pytest_rerunfailures.py:378]: `i` is a hash of the test name (`test_file.py::test_my_test`), `k` is `f` for failures or `r` for reruns. v is the number of failures or reruns (an int). before we run anything in the test (including setup), the worker needs to record in these in the master so that the master can decide whether the test needs to be rerun if it crashes. the worker must communicate this information because the number of reruns can only be computed by collecting tests, which can only be done in the worker. Again, if xdist is not found this this functionality does nothing.

**Comment by xoviat** [comment]
> There is one skipped test: Could it be that this is your newly written test?

The test is skipped because it requires pytest-xdist development branch. What would you like to do about this?

**Comment by xoviat** [comment]
Again, to clarify, this adds new functionality that reruns a test that crashes a worker with pytest-xdist. If a crash occurs without pytest-xdist, there is nothing that can be done, therefore this code is a no-op without pytest-xdist (the next release as of this writing). However, this functionality is important enough to add because it allows a timed-out test (via pytest-timeout) to be rerun (which cannot be currently done).

File: pytest_rerunfailures.py
Comment (xoviat) [reply pytest_rerunfailures.py:378]: To clarify, the StatusDB is an in-memory database stored on the master that records, for each test, the number of failures and the number of reruns. The Status DB is accessed on the worker through a socket connection that binds to a random free port (assigned by the OS). This will allow (eventually) the remote xdist feature to work (when someone requires it).

File: pytest_rerunfailures.py
Comment (icemac) [reply pytest_rerunfailures.py:271]: I think I understood the goal but `StatusDB` initializes a socket … maybe there should be an actual no-op StatusDB and a base class for the classes for client and server which encapsulates the common code for these ones. But a socket and probably some other code should not be needed in the no-op version.

**Comment by icemac** [comment]
> The test is skipped because it requires pytest-xdist development branch. What would you like to do about this?

Probably this PR could use that development branch for now until it is merged/released.

File: pytest_rerunfailures.py
Comment (icemac) [reply pytest_rerunfailures.py:378]: > `i` is a hash of the test name (`test_file.py::test_my_test`), `k` is `f` for failures or `r` for reruns. v is the number of failures or reruns (an int). before we run anything in the test (including setup), the worker needs to record in these in the master so that the master can decide whether the test needs to be rerun if it crashes. the worker must communicate this information because the number of reruns can only be computed by collecting tests, which can only be done in the worker. Again, if xdist is not found this this functionality does nothing.

It would be important to have this explanation in the code.

File: pytest_rerunfailures.py
Comment (xoviat) [reply pytest_rerunfailures.py:271]: Yes, but the socket is not bound to a port. My understanding is that a socket is no-op until it is bound.

**Comment by xoviat** [comment]
@icemac Is this a good summary of required changes?

1. add explanation for set/get.
2. add xdist development branch to tests

**Comment by icemac** [comment]
@xoviat additionally I'd request:

3. fix the linting issues
4. in case of not using xdist make the StatusDB as minimal as possible

**Comment by xoviat** [comment]
Anything else?

**Comment by icemac** [comment]
The linter still fails, see https://github.com/pytest-dev/pytest-rerunfailures/pull/158/checks?check_run_id=2780766837
You can run the linter tests locally by installing `pre-commit`.

**Comment by xoviat** [comment]
@icemac Try again.

**Comment by xoviat** [comment]
cc @icemac

**Comment by xoviat** [comment]
ping @icemac

**Comment by icemac** [comment]
The checks are green now, currently I do not have the time and energy for a final review of this PR. (Maybe @sallner can take over.)
Additionally we need a release of `pytest-xdist` before this PR can be merged.

**Comment by xoviat** [comment]
@icemac Is there an update on this? I'd be willing to resolve the merge conflicts if I could get an indication that this was moving forward.

**Comment by icemac** [comment]
@xoviat Thank you for coming back to this PR. Has the needed xdist release already been created?

**Comment by xoviat** [comment]
Yes

File: .github/workflows/test.yml
Comment (icemac) [inline .github/workflows/test.yml:64]:
```suggestion
          python -m pip install git+https://github.com/pytest-dev/pytest-xdist.git@7f07d50
```

This line should no longer be necessary or at least install a released version.

**Comment by icemac** [comment]
Additionally there is a problem the linter shows up.

**Comment by xoviat** [comment]
@icemac Fixes appear to work on my branch. To be clear, pytest-xdist is not required with this pr, but it is required if you want to run all of the tests (in this repository, not in general).

**Review by icemac — APPROVED** [review]
LGTM.

I will merge it after the CI checks have passed.

**Review by icemac — CHANGES_REQUESTED** [review]
Just before merging  I saw that documentation and a change log entry are missing.

Could you please add them? (in `README.rst` resp. `CHANGES.rst`)

**Comment by xoviat** [comment]
@icemac Done.

**Review by icemac — APPROVED** [review]

**Comment by icemac** [comment]
Thank you for this PR. 😃

**Comment by icemac** [comment]
I just created a release including this PR: https://pypi.org/project/pytest-rerunfailures/10.2/

**Pull request merged**
